types/src/chain_id.rs

```diff
@@ -0,0 +1,49 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use failure::{self, format_err};
+use std::convert::TryFrom;
+
+#[derive(Copy, Clone, Eq, PartialEq, Debug)]
+/// Libra protocols should not be confused by other instances of the Libra
+/// protocol that may be running concurrently. For instance, discovery notes made
+/// on Libra testnet should not be valid on Libra mainnet. A ChainId allows us to
+/// provide domain separation between Libra instances.
+///
+/// Current mappings:
+///
+/// Libra mainnet <=> 1
+/// Libra testnet <=> 10
+pub struct ChainId(u32);
+
+impl ChainId {
+    pub const fn new(id: u32) -> Self {
+        Self(id)
+    }
+
+    /// The chain id for Libra mainnet.
+    pub const fn mainnet() -> Self {
+        Self(1)
```

Comment:
how does this differ from creating top level constants (`lazy static`?) for mainnet and testnet chain ids?

Reply:
`const fn` means it can be evaluated at compile time, so it's effectively a constant. In fact, you could even do

```rust
const chain_id_mainnet: ChainId = ChainId::mainnet();
```

which wouldn't be possible unless `ChainId::mainnet` is a `const fn`. In contrast, `lazy-static` wraps things in a spin lock to ensure they get properly initialized on first access.

Reply:
Assuming that this will eventually become a config parameter, I'm fine either way.